File: VerdelakCDSite/Models/DinoModels.cs
namespace VerdelakCDSite.Models;

public sealed class DinoIndexViewModel
{
    public string? SearchTerm { get; init; }

    public string? TaxonomyFilter { get; init; }

    public string ApiBaseUrl { get; init; } = string.Empty;

    public string? ErrorMessage { get; init; }

    public IReadOnlyList<PublicDinosaurSummary> Dinosaurs { get; init; } = [];

    public IReadOnlyList<PublicDinoTaxonomyNode> Taxonomy { get; init; } = [];

    public int WithImagesCount => Dinosaurs.Count(item => !string.IsNullOrWhiteSpace(item.PrimaryImageUrl));

    public int WithDescriptionsCount => Dinosaurs.Count(item => !string.IsNullOrWhiteSpace(item.Description));

    public bool HasSearch => !string.IsNullOrWhiteSpace(SearchTerm);

    public bool HasTaxonomyFilter => !string.IsNullOrWhiteSpace(TaxonomyFilter);
}

public sealed class DinoDetailViewModel
{
    public PublicDinosaurDetail? Dinosaur { get; init; }

    public string? ErrorMessage { get; init; }
}

public class PublicDinosaurSummary
{
    public int Id { get; init; }

    public string CommonName { get; init; } = string.Empty;

    public string ScientificName { get; init; } = string.Empty;

    public string Slug { get; init; } = string.Empty;

    public IReadOnlyList<PublicDinoClassification> Classification { get; init; } = [];

    public IReadOnlyList<string> Clades { get; init; } = [];

    public string? DiscoveryDate { get; init; }

    public string? DiscoveredBy { get; init; }

    public string? Description { get; init; }

    public string? PrimaryImageUrl { get; init; }

    public string? PrimaryImageCaption { get; init; }

    public string ClassificationLabel => Classification.Count == 0
        ? "Classification not set"
        : string.Join(" / ", Classification.Select(item => $"{item.Rank}: {item.Name}"));
}

public sealed class PublicDinosaurDetail : PublicDinosaurSummary
{
    public IReadOnlyList<PublicDinoContentSection> Sections { get; init; } = [];

    public IReadOnlyList<PublicDinoIllustration> Illustrations { get; init; } = [];
}

public sealed class PublicDinoClassification
{
    public string Rank { get; init; } = string.Empty;

    public string Name { get; init; } = string.Empty;
}

public sealed class PublicDinoContentSection
{
    public int Id { get; init; }

    public string Heading { get; init; } = string.Empty;

    public string Body { get; init; } = string.Empty;

    public int SortOrder { get; init; }
}

public sealed class PublicDinoIllustration
{
    public int Id { get; init; }

    public string ImageUrl { get; init; } = string.Empty;

    public string? Caption { get; init; }

    public string? Credit { get; init; }

    public int SortOrder { get; init; }
}

public sealed class PublicDinoTaxonomyNode
{
    public int Id { get; init; }

    public string Rank { get; init; } = string.Empty;

    public string Name { get; init; } = string.Empty;

    public int? ParentId { get; init; }

    public string? ParentName { get; init; }

    public string? Description { get; init; }

    public int EntryCount { get; init; }
}

File: Verdelak.Angular/src/app/features/dino/models/dino.models.ts
export interface DinoTaxonomyNode {
  id: number;
  rank: string;
  name: string;
  parentId: number | null;
  parentName: string | null;
  description: string | null;
  sortOrder: number;
}

export interface DinoTaxonomyNodeUpsert {
  rank: string;
  name: string;
  parentId: number | null;
  description: string | null;
  sortOrder: number;
}

export interface DinosaurSummary {
  id: number;
  commonName: string;
  scientificName: string;
  slug: string;
  kingdomName: string | null;
  phylumName: string | null;
  className: string | null;
  clades: string | null;
  familyName: string | null;
  subfamilyName: string | null;
  genusName: string | null;
  speciesName: string | null;
  discoveryDate: string | null;
  discoveredBy: string | null;
  hasDescription: boolean;
  sectionCount: number;
  illustrationCount: number;
  isPublished: boolean;
}

export interface DinosaurDetail {
  id: number;
  commonName: string;
  scientificName: string;
  slug: string;
  kingdomId: number | null;
  kingdomName: string | null;
  phylumId: number | null;
  phylumName: string | null;
  classId: number | null;
  className: string | null;
  clades: string | null;
  familyId: number | null;
  familyName: string | null;
  subfamilyId: number | null;
  subfamilyName: string | null;
  genusId: number | null;
  genusName: string | null;
  speciesId: number | null;
  speciesName: string | null;
  discoveryDate: string | null;
  discoveredBy: string | null;
  description: string | null;
  isPublished: boolean;
  sections: DinoContentSection[];
  illustrations: DinoIllustration[];
}

export interface DinosaurUpsert {
  commonName: string;
  scientificName: string;
  slug: string | null;
  kingdomId: number | null;
  phylumId: number | null;
  classId: number | null;
  clades: string | null;
  familyId: number | null;
  subfamilyId: number | null;
  genusId: number | null;
  speciesId: number | null;
  discoveryDate: string | null;
  discoveredBy: string | null;
  description: string | null;
  isPublished: boolean;
  sections: DinoContentSectionUpsert[];
  illustrations: DinoIllustrationUpsert[];
}

export interface DinoContentSection {
  id: number;
  heading: string;
  body: string;
  sortOrder: number;
}

export interface DinoContentSectionUpsert {
  id: number | null;
  heading: string;
  body: string;
  sortOrder: number;
}

export interface DinoIllustration {
  id: number;
  imageUrl: string;
  caption: string | null;
  credit: string | null;
  sortOrder: number;
}

export interface DinoIllustrationUpsert {
  id: number | null;
  imageUrl: string;
  caption: string | null;
  credit: string | null;
  sortOrder: number;
}

export interface PublicDinoClassification {
  rank: string;
  name: string;
}

export interface PublicDinosaurSummary {
  id: number;
  commonName: string;
  scientificName: string;
  slug: string;
  classification: PublicDinoClassification[];
  clades: string[];
  discoveryDate: string | null;
  discoveredBy: string | null;
  description: string | null;
  primaryImageUrl: string | null;
  primaryImageCaption: string | null;
}

export interface PublicDinosaurDetail extends PublicDinosaurSummary {
  sections: DinoContentSection[];
  illustrations: DinoIllustration[];
}

export interface PublicDinoTaxonomyNode {
  id: number;
  rank: string;
  name: string;
  parentId: number | null;
  parentName: string | null;
  description: string | null;
  entryCount: number;
}
